src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -198,27 +175,32 @@ public IJsonTypeInfoResolver TypeInfoResolver
             [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
             get
             {
-                if (_typeInfoResolver != null)
-                {
-                    return _typeInfoResolver;
-                }
-
-                EnsureInitializedForReflectionSerializer();
-                return _typeInfoResolver;
+                return _typeInfoResolver ?? DefaultJsonTypeInfoResolver.DefaultInstance;
             }
             set
             {
                 VerifyMutable();
 
-                if (value == null)
+                if (value is null)
                 {
                     throw new ArgumentNullException(nameof(value));
                 }
 
+                if (value is JsonSerializerContext ctx)
+                {
+                    if (ctx._options != null && ctx._options != this)
+                    {
+                        // TODO evaluate if this is the appropriate behaviour;
+                        // users are still able to work around this check by wrapping JsonTypeInfoResolver.Combine(value);
+                        ThrowHelper.ThrowInvalidOperationException_SerializerContextOptionsImmutable();
+                    }
+
+                    // Associate options instance with context and lock for further modification
+                    ctx._options = this;
+                    _isLockedInstance = true;
+                }
+
```

Comment:
we don't need to clear effective resolver here because that's only possible for context which won't allow changing it after it's set?

Reply:
Yeah, it should only be set once the instance is locked for furher modification.